pest/src/inputs/string_input.rs

```diff
@@ -30,10 +36,182 @@ impl StringInput {
     /// assert_eq!(input.len(), 3);
     /// ```
     pub fn new(string: String) -> StringInput {
-        StringInput {
-            string
+        StringInput { string }
+    }
+}
+
+impl<'a> StrInput<'a> {
+    /// Creates a new `StrInput` from a `&str`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use pest::inputs::{Input, StrInput};
+    /// let input = StrInput::new("asd");
+    ///
+    /// assert_eq!(input.len(), 3);
+    /// ```
+    pub fn new(source: &'a str) -> StrInput<'a> {
+        StrInput { str_ref: source }
+    }
+}
+
+#[inline]
+unsafe fn line_col(source: &str, pos: usize) -> (usize, usize) {
+    if pos > source.len() {
+        panic!("position out of bounds");
+    }
+
+    let mut pos = pos;
+    let slice = &source[..pos];
+    let mut chars = slice.chars().peekable();
+
+    let mut line_col = (1, 1);
+
+    while pos != 0 {
+        match chars.next() {
+            Some('\r') => {
+                if let Some(&'\n') = chars.peek() {
+                    chars.next();
+
+                    if pos == 1 {
+                        pos -= 1;
+                    } else {
+                        pos -= 2;
+                    }
+
+                    line_col = (line_col.0 + 1, 1);
+                } else {
+                    pos -= 1;
+                    line_col = (line_col.0, line_col.1 + 1);
+                }
+            }
+            Some('\n') => {
+                pos -= 1;
+                line_col = (line_col.0 + 1, 1);
+            }
+            Some(c) => {
+                pos -= c.len_utf8();
+                line_col = (line_col.0, line_col.1 + 1);
+            }
+            None => unreachable!(),
```

Comment:
There should be no comma here. Hm, newest version ofrustfmt should have dealt with this particular case.